Lead Details Page: initial page setup

blakelester opened the pull request:
This is mostly just a code check-in, not really going for full functionality at the moment (PR would be much bigger).

Changes
1. Setup Lead Details page (work in progress): added some basic form fields and validation
2. Modified nav bar to remove separate "home" link
3. added "unit numbers" column to leads table
4. updated to work with new `/leads` endpoint
5. removed some status type cards from home page per feedback

<img width="1280" alt="Screen Shot 2022-02-10 at 11 27 32 PM" src="https://user-images.githubusercontent.com/55535606/153546515-033e5512-fd62-4330-b821-bf3b9a45da35.png">

Commit:
checking in initial code for lead details page

Commit:
fixed typo in lead details endpoint

Commit:
additional work on Roach Assessment form

Ashley-Owens commented on src/components/forms/LeadFormRoachAssessment.js:
```diff
@@ -0,0 +1,270 @@
+        if ('zip' in fieldValues) {
+            temp.zip = fieldValues.zip ? "" : "This field is required.";
+        }
+        if ('buildingNums' in fieldValues) {
```
Can you make the Service Details a separate form? We need 25 different ones so will be easier to make them modular rather than having one large if statement block.

Ashley-Owens commented on src/components/forms/LeadFormRoachAssessment.js:
```diff
@@ -0,0 +1,270 @@
+                <Grid container>
+                    <Grid item xs={4}>
+                        <Controls.Input
+                            name="service"
```
Make Service form a separate component as above...

Ashley-Owens reviewed (commented):
Hey Blake, this is looking great. Your PR didn't mention To Do items so forgive me if you already know to change things... First, I really like the data presented in the form, that looks awesome. Edit feature is updating in the database, nice job on that.

When I go to edit the form it's saying some fields are required that aren't (notes, contact) and some fields aren't required that are (target pest, service):
![Screen Shot 2022-02-14 at 8 27 42 AM](https://user-images.githubusercontent.com/54197270/153894219-2baddfa6-40dc-4b35-837d-5c65bb5b6000.png)

We need this status to be editable between New, Closed, and Open.
![Screen Shot 2022-02-14 at 8 33 03 AM](https://user-images.githubusercontent.com/54197270/153894358-74b24c35-b5d3-40f5-b0a6-df434b8ebc58.png)

When the user clicks to Edit this form, would it be feasible to change the Edit button to say Cancel? That would add some functionality to the button while in the edit screen.
![Screen Shot 2022-02-14 at 8 35 55 AM](https://user-images.githubusercontent.com/54197270/153894915-e09cc507-fcc9-4265-b119-55bcd5d13adb.png)

We've got an email chain going that has the Smart Pest logo. When you get a chance can you update it?
![Screen Shot 2022-02-14 at 8 41 32 AM](https://user-images.githubusercontent.com/54197270/153896204-519ee2ba-ca7b-46b3-afde-b2ec2cc09b27.png)

Also, looks like you're doing another Get request when the user selects an item, even though you've already pulled that data. Curious why we don't just access the data you've already received to build out the forms?

Looks like you've combined the general Lead form info with Roach Assessment service. Could you make Roach a separate component? Just a suggestion but I think it would be easier to make your Form Components based off Pest types like Roach (now has 4 different services), bed bug, bee & wasp, etc. That will save you from having to make 25 different forms for all the different services and will help eliminate redundant code when General info is combined with each Service/Pest type.

Overall it's coming along great. Hopefully things will go faster once you get one form finished. Let me know if you have any questions on the feedback..
Thanks!

blakelester commented:
@Ashley-Owens Thanks for the feedback, I should have listed out the more of the "to do" items as it's still in a very rough state. I agree with all your feedback and have most of that on my running to do list. Some good call outs that I didn't have yet:

1. status should be an editable field
2. validation, but totally missed that some of those fields are optional and others that are required haven't marked as such
3. edit/save button: I'm not happy with the current state. I was thinking about making "edit" the only action button that appears on first page load. When you click edit, the primary action changes to "save". Perhaps also having a cancel button to the left of it. I still need to devote some time to this though. Officially adding that to the list!

Re: splitting the form components up. The form was giving me some much trouble with managing state... I couldn't figure out how to split those up yet and yet keep the input fields' state in sync. I'll have to circle back to that before I get ready to work on the next form. Agree, that would be so nice to have.

Commit:
addressed some feedback

Commit:
split General and Service form field components

blakelester commented:
Additional work completed:

1. Split up "General" and "Service" form fields into their own components. This will make it much more "DRY" and maintainable.
2. Polished the Edit, Save, Cancel button behavior
3. Added Smart Pest logo to nav bar
4. Made status an editable form field
5. Still doing a GET request for individual lead when loading the Details page, but this is intentional. I've seen this behavior in other web apps.

Note: validation isn't pretty and should be organized better, but I've been beating my head against a wall all day with validation bugs, so I'm going to write the validation rules in the `validation` function in the `LeadForm` component for now.

Ashley-Owens reviewed (commented):
Hey Blake,
This is coming along nicely. I really like the button behavior, it's very cool.

I'm not sure if you've got validation working but here it says that the tech Made Contact, yet the contact info stuff is empty and it should be requiring me to fill it with data. Maybe in testing it was accidentally saved as null because mobile is requiring it when Made Contact is true.
![Screen Shot 2022-02-21 at 4 55 23 PM](https://user-images.githubusercontent.com/54197270/155039948-77d5ea94-bb04-43a7-a237-a7dd5ce068be.png)
![Screen Shot 2022-02-21 at 4 57 01 PM](https://user-images.githubusercontent.com/54197270/155040026-6bad7668-32e8-4fff-84d1-b249fb9942f7.png)

Could you make that status a drop down since it can only be open, new or closed? That will prevent erroneous data from being saved in this spot. Otherwise, I'm getting some warnings when running it but everything seems to be working ok.

Commit:
address Ashley's feedback

Commit:
update leads endpoints for new and open

Ashley-Owens reviewed (commented):
Ok,
Drop down is much better, can we remove None from both drop downs?
Now that I can see some data in the contactTitle box, can you make this a drop down to mirror the selections Ryan requested for this field? Ellen has a pic on Teams.
Validation is working, thanks so much for the fix.

Commit:
more feedback from ashley

Ashley-Owens reviewed (approved):
Looking good!
